power_grid_model_c/power_grid_model/include/power_grid_model/component/component.hpp

```diff
@@ -0,0 +1,39 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "../common/common.hpp"
+#include "../common/enum.hpp"
+
+#include <concepts>
+#include <string_view>
+
+namespace power_grid_model {
+
+// change of update cause topology and param change, or just param change
+struct UpdateChange {
+    bool topo{};
+    bool param{};
+
+    friend constexpr UpdateChange operator||(UpdateChange const& x, UpdateChange const& y) {
+        return UpdateChange{x.topo || y.topo, x.param || y.param};
+    }
+};
+
+template <typename T>
+concept component_c = requires(T t, T const& ct, typename T::UpdateType u, typename T::UpdateType const& cu) {
+                          typename T::InputType;
+                          typename T::UpdateType;
+
+                          { T::name } -> std::convertible_to<std::string_view>;
+                          { ct.math_model_type() } -> std::convertible_to<ComponentType>;
```

Comment:
In the state queries, I see
```
namespace power_grid_model::main_core {
template <typename ComponentType, class ComponentContainer>
    requires model_component_state_c<MainModelState, ComponentContainer, ComponentType>
inline Idx get_component_type_index(MainModelState<ComponentContainer> const& state) {
    return state.components.template get_type_idx<ComponentType>();
}
```
and also in main core:
```
template <template <typename T> class StateType, typename ContainerType, typename ComponentType>
concept model_component_state_c =
    component_container_c<typename StateType<ContainerType>::ComponentContainer, ComponentType> &&
    std::same_as<StateType<ContainerType>, MainModelState<ContainerType>>;
```

Is it intentional for the type name and enum to be the same?

Reply:
not intentional but also not harmful because only one of the two is used within the template functions